lib/puppet/transaction/report.rb

```diff
@@ -323,7 +333,7 @@ def to_data_hash
       'environment' => @environment,
       'logs' => @logs.map { |log| log.to_data_hash },
       'metrics' => Hash[@metrics.map { |key, metric| [key, metric.to_data_hash] }],
-      'resource_statuses' => Hash[@resource_statuses.map { |key, rs| [key, rs.nil? ? nil : rs.to_data_hash] }],
+      'resource_statuses' => Hash[@resource_statuses.map { |key, rs| [key, skip_or_to_data_hash(rs)] }],
```

Comment:
If we serialize a resource status with a `nil` entry like:

```
 "resource_statuses"=>
  {"File[/tmp/pass]"=>nil,
```
Then puppetserver will barf trying to process the report:
```
2022-11-01T23:07:22.887Z INFO  [qtp530249092-53] [puppetserver] Puppet Bad Request: The request body is invalid: Could not intern from json: undefined method `[]' for nil:NilClass
```

But if we serialize an empty hash, then it works:
```
"resource_statuses"=>
  {"File[/tmp/pass]"=>{},
```

I'm guessing `rs.nil?` is only ever true in tests? So maybe we should serialize `Puppet::Resource::EMPTY_HASH` instead of `nil`?

Reply:
Does it make sense to introduce new static map so that there is a clue why it's empty? Something like `Puppet::Resource::UNCHANGED_RESOURCE` where it's just a map with a simple message stating the resource is unchanged?

Reply:
Ah one thing I just noticed, previously `rs.to_data_hash` always returned a new mutable hash, but if we use the frozen `EMPTY_HASH`, then the resource names will refer to the same object:

```
irb(main):003:0> resource_statuses = {1 => 2, 3 => 4}
=> {1=>2, 3=>4}
irb(main):004:0> EMPTY = {}.freeze
=> {}
irb(main):005:0> h = Hash[resource_statuses.map { |k, v| [k, EMPTY] }]
=> {1=>{}, 3=>{}}
irb(main):006:0> h[1].object_id
=> 180
irb(main):007:0> h[3].object_id
=> 180
```

which could cause issues if someone is relying on being able to modify the resource status objects. To preserve the original behavior, I'd recommend returning a new hash:

```ruby
elsif !rs.out_of_sync? && Puppet[:exclude_unchanged_resources]
  {}
```